# Assignment 5

These assignments are related to **Linked List**.
* Read the related course material before doing the assignments from the [Linked List](https://ttc1050.pages.labranet.jamk.fi/ds_alg_material/materials/05_linked_list/).

General notes of assignments:
* NOTE! In general, after the implementation of the function, all assignments have a test program for the function.
* NOTE! Add your code in the assignments only after the TODO lines.

## Assignment 5.1 (1 p)

Report the functionality of the program.

Here is the implementation of an algorithm that performs some _linked list operations_.

Please
1. _Analyze the code_ and
2. _Report on the functionality_ of this program.

Note! You should use the debugger with breakpoints there to test the program.

In [16]:
class Point:
    def __init__(self, value=None, next=None):
        self.value = value
        self.next = None


class List:
    def __init__(self):
        self.head = None

    def operation_1(self, numb):
        while numb > 1:
            numb = self.operation_2(numb)
            p = Point(numb)

            if self.head is None:
                self.head = p
            else:
                cur = self.head

                while cur.next:
                    cur = cur.next

                cur.next = p

        c = self.head
        n = []
        while c:
            n.append(str(c.value))
            c = c.next
        print(' -> '.join(n))

    def operation_2(self, x):
        y = x / 2
        return y


new_l = List()
num = 100
new_l.operation_1(num)

50.0 -> 25.0 -> 12.5 -> 6.25 -> 3.125 -> 1.5625 -> 0.78125


### Your response

### Class Definitions

#### Point Class
- It is an element of the linked list.
- Contains two attributes: `value` to store the data and `next`= None to point to the next element in the list.

#### List Class
- It is the linked list.
- Contains a single attribute `head` = None , which points to the first element of the linked list.

### List Class Methods

#### `operation_1(self, numb)`
- This method takes a number `numb` as an argument and performs operations on it until `numb` becomes less than or equal to 1.
- It continuously cut `numb` in half by calling `operation_2` and creates a new `Point` instance with the halved value.
- This new point is then added to the end of the linked list using append.
- After the loop is done (when `numb` becomes less than or equal to 1), it traverses the linked list from the head, collects the values of each node, and prints them joined by " -> ".

#### `operation_2(self, x)`
- A simple method that takes a number `x` and returns divides it by 2 and return the result (`x / 2`).

### Main Program Flow

1. A `List` object `new_l` is created.
2. The number 50 is assigned to `num`.
3. `operation_1` is called with `num` as its argument.

### What the Program Does

1. Starts with `numb = 50`.
2. Halves `numb` in each iteration of the `while` loop in `operation_1`, creating and adding a new `Point` with the halved value to the linked list each time. This process continues until `numb` becomes less than or equal to 1.
3. For `numb = 50`, the sequence of halved values added to the list would be: 25, 12.5, 6.25, 3.125, 1.5625.
4. Once `numb` is less than or equal to 1, the loop stops, and the program prints the values in the linked list, which would be "25 -> 12.5 -> 6.25 -> 3.125 -> 1.5625".

### Conclusion

The program demonstrates a simple linked list operation where elements are added to the end of the list. Each element represents a halved value starting from an initial number (50 in this case) until the value becomes less than or equal to 1. This showcases basic linked list manipulation, including traversal and insertion at the end.

## Assignment 5.2 (1 p)

Implement new methods to `LinkedList` class.

This Python code defines two classes: `Node` and `LinkedList`, implementing a basic **linked list** structure.
* `Node` represents a node in the linked list and
* `LinkedList` represents the linked list itself.

In this assignment `LinkedList` implementation allows creating a linked list, add nodes, display the contents, and performing an iterative search for a specific value.

Do the following task:
* The `LinkedList` class will store the following integers: `3, 31, 21, 52, 16, 91, 10, and 5`. Insert these integers into the linked list.
* Implement a search method called `iterative_search()` in the `LinkedList` class that takes the *Integer* to be searched for as an argument and reports whether this integer is found in the `LinkedList`.
* Implement `search_items_from_list()`calling `iterative_search()` method from LinkedList class.
* In test, you must add all items from `added_values` list to `linked_list`.

In [15]:
correct_answer_05_02 = [True, True, False]

added_values = [13, 31, 21, 52, 16, 91, 10, 5]

import unittest

# Class for linkedlist node
class Node:
    def __init__(self, data=None, next_item=None):
        self.data = data
        self.next = next_item


# Class for linkedlist
class LinkedList:
    def __init__(self):
        self.head = None

    def add_node(self, data):
        new_node = Node(data)
        if self.head:
            current = self.head
            while current.next:
                current = current.next
            current.next = new_node
        else:
            self.head = new_node
    def display(self):
        current = self.head
        while current:
            print(current.data)
            current = current.next

    def iterative_search(self, n):
        current = self.head
        while current:
            if current.data == n:
                return True
            current = current.next
        return False

def search_items_from_list(to_find, llist):
    results = []
    for item in to_find:
        result = llist.iterative_search(item)
        results.append(result)
    return results

# Test
added_values = [13, 31, 21, 52, 16, 91, 10, 5]
linked_list = LinkedList()

# TODO: Add all items from added_values to LinkedList.
for value in added_values:
    linked_list.add_node(value)

linked_list.display()

search_items = [5, 31, 22]
found_items = search_items_from_list(search_items, linked_list)
print(found_items)

unittest.TestCase().assertListEqual(found_items, correct_answer_05_02)
print(f"answer {found_items} was OK!")

13
31
21
52
16
91
10
5
[True, True, False]
answer [True, True, False] was OK!


## Assignment 5.3 (1 p)

Sort linked list elements.

Create a function called `ask_linked_list_from_user()` that prompts the user to input the number of nodes in the linked list.
Subsequently, ask the user to enter the same number of strings to be saved into the linked list.

Perform also the following operations:
* Sort the order of the linked list using the bubble sort method. You must implement bubble sort in this assignment in `bubble_sort_for_linked_list()` function
* Create a method that returns the content of the linked list as a `list`.

Note! The test program has already been implemented, so please don't edit it.

In [14]:
import unittest
class Node:
    def __init__(self, data=None, next_item=None):
        self.data = data
        self.next = next_item


class LinkedList:
    def __init__(self):
        self.head = None

    def add_node(self, data):
        new_node = Node(data)
        if self.head:
            current = self.head
            while current.next:
                current = current.next
            current.next = new_node
        else:
            self.head = new_node

    def display(self):
        current = self.head
        while current is not None:
            if current.data is not None:
                print(current.data)
            current = current.next

    def to_list(self):
        new_list = []
        current = self.head
        while current is not None:
            if current.data is not None:
                new_list.append(current.data)
            current = current.next
        return new_list

def iterative_search(self, n):
    current = self.head
    while current:
        if current.data == n:
            return True
        current = current.next
    return False


def bubble_sort_for_linked_list(llist):
    end = None
    while end != llist.head:
        p = llist.head
        while p.next != end:
            q = p.next
            if p.data > q.data:
                p.data, q.data = q.data, p.data
            p = p.next
        end = p

def ask_linked_list_from_user():
    llist = LinkedList()
    num_nodes = int(input("Enter the number of nodes: "))
    for _ in range(num_nodes):
        data = input("Enter node value: ")
        llist.add_node(data)
    return llist


# Test
linked_list = ask_linked_list_from_user()
sorted_orig_list = sorted(linked_list.to_list())

bubble_sort_for_linked_list(linked_list)
sorted_llist = linked_list.to_list()

unittest.TestCase().assertListEqual(sorted_llist, sorted_orig_list)
print(f"answer {sorted_llist} was OK!")


Enter the number of nodes:  5
Enter node value:  2
Enter node value:  3
Enter node value:  5
Enter node value:  6
Enter node value:  8


answer ['2', '3', '5', '6', '8'] was OK!


## Assignment 5.4

Create **Double Linked List** that stores `Person` objects.

In a Doubly linked list, each node contains a data element and two pointers, one pointing to the next node in the sequence and another pointing to the previous node.
It allows for traversal in both directions, forward and backward, to perform operations such as insertion and deletion at both ends of the list.

You should create the following methods in the `DoubleLinkedList` class:
* `add_node()`: Add a new `Node` to Double linked list.
* `Insert_at_the_beginning()`: Insert a new `Node` at the beginning of Double linked list.
* `remove_node()`: Optional part is to remove a `Node` from the Double linked list. It's optional because the test program does not call this method.

Note! The test program has already been implemented, so please don't edit it.

In [13]:
# Class for linkedlist node
class Node:
    def __init__(self, data=None, next=None):
        self.prev = None
        self.data = data
        self.next = None


# Class for linkedlist
class DoubleLinkedList:
    def __init__(self):
        self.head = None

    def add_node(self, data):
        new_node = Node(data)
        if self.head is None:
            self.head = new_node
        else:
            current = self.head
            while current.next:
                current = current.next
            current.next = new_node
            new_node.prev = current
        self.display()

    def insert_at_the_beginning(self, data):
        new_node = Node(data)
        if self.head is not None:
            self.head.prev = new_node
            new_node.next = self.head
        self.head = new_node
        self.display()

    def remove_node(self, target):
        current = self.head
        while current:
            if current.data.name == target.name and current.data.age == target.age:
                if current == self.head:
                    self.head = current.next
                    if self.head:
                        self.head.prev = None
                else:
                    if current.next:
                        current.next.prev = current.prev
                    if current.prev:
                        current.prev.next = current.next
                print(f"Removed {target}")
                self.display()
                return
            current = current.next
        print(f"Element {target} does not exist in the list.")

    def display(self):
        current = self.head
        nodes = []
        while current:
            nodes.append(str(current.data))
            current = current.next
        print(' <-> '.join(nodes))

    def to_string(self):
        current = self.head
        nodes = []
        while current:
            nodes.append(f"{current.data.name}, {current.data.age}")
            current = current.next
        return ' : '.join(nodes)



class Person:
    def __init__(self, name, age):
        self.name = name
        self.age = age

    def name(self):
        return self.name

    def age(self):
        return self.age

    def __str__(self):
        return f"{self.name}, {self.age}"


# Test
llist = DoubleLinkedList()

llist.add_node(Person("Bob", 45))
llist.add_node(Person("Jill", 37))
llist.add_node(Person("Jane", 29))
llist.insert_at_the_beginning(Person("Jack", 59))
# Optional part
llist.remove_node(Person("Jack", 59))

llist.display()

result = "Bob, 45 : Jill, 37 : Jane, 29" 

assert llist.to_string() == result

print(f"answer '{llist.to_string()}' was OK!")

Bob, 45
Bob, 45 <-> Jill, 37
Bob, 45 <-> Jill, 37 <-> Jane, 29
Jack, 59 <-> Bob, 45 <-> Jill, 37 <-> Jane, 29
Removed Jack, 59
Bob, 45 <-> Jill, 37 <-> Jane, 29
Bob, 45 <-> Jill, 37 <-> Jane, 29
answer 'Bob, 45 : Jill, 37 : Jane, 29' was OK!
